### PR TITLE
EIP-0027: Emission Soft-Fork

### PR DESCRIPTION
**Pull request by kushti**
Long-term security of the Ergo protocol, including its crypto-economic security, is always of highest priority for the community 
and core developers. One of the hottest research topic in this field is possible (in)stability of cryptocurrency protocols without stable block rewards coming from emission. 
 
It was planned during the launch of the Ergo network that after the end of emission miners will be rewarded with 
transaction fees and also with storage rent (unique for Ergo source of mining income). However, it is hard to estimate yet how successfully will storage rent replace emission. 

Thus it was proposed on the ErgoForum informally to prolong emission (with preserving total supply) 
via a soft-fork. This EIP is concluding previous discussions and provides details on emission soft-fork 
design and implementation.

### CODE REVIEW

File: eip-0027.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+Miners are enforced to pay to this proxy contract. The contract is similar to the existing fee contract but with no time lock. 
+
+* We create new consensus-level checks that verify that the proper amount of re-emission tokens are going from the emission box to a miner rewards box, and also that during spending a box which is not an emission box but contains re-emission tokens (miner rewards box), the tokens are being burnt, and the same amount of nanoERG is locked by the re-emission contract. 
+
+The checks can be switched off via a soft-fork (via voting on disabling rule #123).
+
+Contracts
+-------------------
+
+**Re-emission contract**: This contract pays 3 ERG per block starting from block #2,080,800 . Also, this contract allows for merging with other boxes; the merging transaction must have only two outputs; the first output is for the re-emission contract, the second one is to pay a mining fee (its value can be 0.01 ERG at most)
```
Comment (aslesarenko) [inline eip-0027.md]: Please describe merging part in more details. Why it is called `sponsored` in the contract?

**Review by aslesarenko — CHANGES_REQUESTED** [review]
Also please see #66

File: eip-0027.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+* The amount of re-emission tokens in a mining rewards box shows how many ERG a miner should send to the re-emission contract when spending the box. Any miner can do the injection, which happens on activation height.
+
+We introduce two new contracts which are described in greater detail below. 
+* The **re-emission contract** pays 3 ERG for each block after the end of emission.
+
+* The **pay-to-reemission** contract, is a proxy contract active before the re-emission contract. 
+
+Miners are enforced to pay to this proxy contract. The contract is similar to the existing fee contract but with no time lock. 
+
+* We create new consensus-level checks that verify that the proper amount of re-emission tokens are going from the emission box to a miner rewards box, and also that during spending a box which is not an emission box but contains re-emission tokens (miner rewards box), the tokens are being burnt, and the same amount of nanoERG is locked by the re-emission contract.
```
Comment (greenhat) [inline eip-0027.md]: Could you please elaborate on these new consensus-level checks? I kinda get the big picture and checked the re-emission contract for possible ways to steal the funds. But I'm struggling to understand the exact mechanism of re-emission token flow from the emission box to the miner's box and how the miner "sends" ERGs to the re-emission contract. I think a step-by-step walkthrough of the miner's reward path (emission->miner->re-emission) with links to the source code of the consensus-level checks might explain it all.

**Review by greenhat — CHANGES_REQUESTED** [review]
I think I get the big picture, but I'm struggling to understand some parts of the workflow. Please see my comment on the new consensus-level checks.

**Comment by aslesarenko** [comment]
@scalahub I would wait your review before approving this, could you please have a look.

File: eip-0027.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+    val firstTokenId = SelectField(ByIndex(rOutTokens, IntConstant(0)), 0.toByte)
+
+    val correctNftId = EQ(firstTokenId, ByteArrayConstant(reemissionNftId))
+    
+    // output of the reemission contract
+    val reemissionOut = ByIndex(Outputs, IntConstant(0))
+    
+    // output to pay miner
+    val minerOut = ByIndex(Outputs, IntConstant(1))
```
Comment (scalahub) [inline eip-0027.md:96]: Already defined in line 79

File: eip-0027.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+    val rOutTokens = OptionGet(ExtractRegisterAs(reemissionOut, R2)(SCollection(STuple(SCollection(SByte), SLong))))
+
+    val firstTokenId = SelectField(ByIndex(rOutTokens, IntConstant(0)), 0.toByte)
+
+    val correctNftId = EQ(firstTokenId, ByteArrayConstant(reemissionNftId))
+    
+    // output of the reemission contract
+    val reemissionOut = ByIndex(Outputs, IntConstant(0))
```
Comment (scalahub) [inline eip-0027.md:93]: Already defined in line 76

**Comment by scalahub** [comment]
For analysis, made an approximate ergoscript version
reemission: https://gist.github.com/scalahub/ff5d62918389e164a5f0af54eab8151a
pay-to-reemission: https://gist.github.com/scalahub/c89a2ab513d1ca26eafe1681a4667316

File: eip-0027.md (hunk 2 above)
Comment (scalahub) [inline eip-0027.md]: > and the same amount of nanoERG is locked by the re-emission contract.

It should be pay-to-re-emission contract. Locking in re-emission contract will become insecure. We have to assume only one copy of re-emission contract exists (see comment below)

File: eip-0027.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        correctNftId,
+        sameScriptRule,
+        OR(
+          sponsored,
+          AND(
+            heightCorrect,
+            correctMinerOutput,
+            afterEmission,
+            heightIncreased,
+            correctCoinsIssued
```
Comment (scalahub) [inline eip-0027.md:137]: I think we also need to ensure one of two things 
A) INPUTS.size == 1
B) When miner spends the reward box, the ergs to re-emit always go to proxy contract and not the re-emission contract (see above comment). 

if we don't ensure both, then multiple re-emission contract boxes can be spent during reward collection

**Review by scalahub — CHANGES_REQUESTED** [review]

File: eip-0027.md (hunk 4 above)
Comment (kushti) [reply eip-0027.md:93]: Duplicate removed

File: eip-0027.md (hunk 3 above)
Comment (kushti) [reply eip-0027.md:96]: Duplicate removed

File: eip-0027.md (hunk 5 above)
Comment (kushti) [reply eip-0027.md:137]: There's only one re-emission contract box collecting rewards from emission contract (which is enforced by NFT condition in pay-to-reemission contract). 

```
INPUTS.size == 1
```

is not needed, better to leave possibility to have additional inputs (e.g. rewarding miners with custom tokens, if first input has re-emission contract box NFT).

File: eip-0027.md (hunk 1 above)
Comment (kushti) [reply eip-0027.md]: Renamed to `merging`, what should be added into description?

Comment (kushti) [reply eip-0027.md]: Renamed to `merging`, what should be added into description?

**Comment by kushti** [comment]
> For analysis, made an approximate ergoscript version reemission: https://gist.github.com/scalahub/ff5d62918389e164a5f0af54eab8151a pay-to-reemission: https://gist.github.com/scalahub/c89a2ab513d1ca26eafe1681a4667316

In re-emission contract:

```
val correctCoinsIssued = SELF.value - reemissionOut.value == 3 
```

should be 3 billion (it is in nanoERG)

File: eip-0027.md (hunk 2 above)
Comment (kushti) [reply eip-0027.md]: Made a note on that

**Review by aslesarenko — APPROVED** [review]

**Review by greenhat — APPROVED** [review]

File: eip-0027.md [hunk 6]
```diff
@@ -60,7 +60,11 @@ We introduce two new contracts which are described in greater detail below.
 
 Miners are enforced to pay to this proxy contract. The contract is similar to the existing fee contract but with no time lock. 
 
-* We create new consensus-level checks that verify that the proper amount of re-emission tokens are going from the emission box to a miner rewards box, and also that during spending a box which is not an emission box but contains re-emission tokens (miner rewards box), the tokens are being burnt, and the same amount of nanoERG is locked by the re-emission contract. 
+* We create new consensus-level checks that verify that the proper amount of re-emission tokens are going from 
+the emission box to a miner rewards box, and also that during spending a box which 
+is not an emission box but contains re-emission tokens (miner rewards box), the tokens are being burnt, 
+and the same amount of nanoERG is locked by the re-emission contract (via a proxy pay-to-reemission contract,
+ ERGs locked by such a contract could be paid to reemission contract box (its NFT) only).
```
Comment (scalahub) [inline eip-0027.md]: This wording is confusing. I think we can simply replace above paragraph with the below (change in bold -- the rest is automatically implied by the contract):

> the emission box to a miner rewards box, and also that during spending a box which 
is not an emission box but contains re-emission tokens (miner rewards box), the tokens are being burnt, 
and the same amount of nanoERG is locked **in a box with the pay-to-reemission contract (aka the proxy contract)**.

**Review by scalahub — CHANGES_REQUESTED** [review]

File: eip-0027.md (hunk 6 above)
Comment (glasgowm148) [reply eip-0027.md]: updated

**Review by glasgowm148 — APPROVED** [review]

**Review by scalahub — APPROVED** [review]

**Comment by scalahub** [comment]
> > For analysis, made an approximate ergoscript version reemission: https://gist.github.com/scalahub/ff5d62918389e164a5f0af54eab8151a pay-to-reemission: https://gist.github.com/scalahub/c89a2ab513d1ca26eafe1681a4667316
> 
> In re-emission contract:
> 
> ```
> val correctCoinsIssued = SELF.value - reemissionOut.value == 3 
> ```
> 
> should be 3 billion (it is in nanoERG)

fixed the gist

**Pull request merged**